lib/graphs.ts

```diff
@@ -171,49 +171,72 @@ export const getTrendAfterIndex = (
 };
 
 /**
- * Determines the moving average, by calculating the averages for a given period in each array
- * @param array Array of GenerationDataPoints to determine the Simple moving averages for
- * @param period the length of each interval of the average, the length of the period must be odd or it will throw an exception
- * @returns a GenerationDataPoint array of Simple Moving Averages for each indexin O(n * p) time unfortunately
+ * Calculate the centered moving average for a given period in O(n) time.
+ *
+ * @param points - the array of points to calculate the centered moving average for
+ * @param period - the period (maximum window size) to calculate the centered moving average for
+ * @returns an array of points with the centered moving average calculated
+ *
+ * @throws an error if the period is even or greater than the length of the array
  */
-
-export const SimpleMovingAverage = (
-  array: GenerationDataPoint[],
+export const calculateCenteredMovingAverage = (
+  points: GenerationDataPoint[],
   period: number
 ): GenerationDataPoint[] => {
   if (period % 2 == 0) {
     throw new Error('Period must be an odd number');
-  } else if (period > array.length) {
+  } else if (period > points.length) {
     throw new Error('Period must be less than or equal to array size');
   }
 
-  let averages: number[] = [];
-
-  const upperBound = Math.floor(period / 2);
+  const averagedPoints: GenerationDataPoint[] = [];
 
-  for (let i = 0; i < array.length; i++) {
-    const valsBefore = i;
-    const valsAfter = array.length - i - 1;
+  // Calculate the prefix sum of the generation data
+  const prefixSum = points.reduce(
+    (acc, curr) => {
+      acc.push(acc[acc.length - 1] + curr.generation_kw);
+      return acc;
+    },
+    [0]
+  );
 
-    const range = Math.min(valsBefore, valsAfter, upperBound);
-    let currSum = 0;
-    for (let j = i - range; j < i + range + 1; j++) {
-      currSum += array[j].generation_kw;
+  // Calculate the centered moving average for each point
+  for (let i = 0; i < averagedPoints.length; i++) {
+    if (i == 0 || i == averagedPoints.length - 1) {
```

Comment:
Is this supposed to be this? It's returning empty rn
```suggestion
  for (let i = 0; i < points.length; i++) {
    if (i == 0 || i == points.length - 1) {
```